implementations/src/main/java/ru/ifmo/nds/jfb/AbstractJFBSorting.java

```diff
@@ -353,6 +353,22 @@ private int helperBMain(int goodFrom, int goodUntil, int weakFrom, int weakUntil
         return splitMerge.mergeTwo(indices, tempFrom, weakFrom, newWeakMidL, weakMidL, newWeakUntil);
     }
 
+    double[][] getPoints(int from, int until, int k) {
+        final double[][] ps = new double[until - from][k];
+        for (int i = from; i < until; i++) {
+            System.arraycopy(this.points[indices[i]], 0, ps[i - from], 0, k);
+        }
+        return ps;
+    }
+
+    int[] getRanks(int from, int until) {
+        final int[] rs = new int[until - from];
+        for (int i = from; i < until; i++) {
+            rs[i - from] = this.ranks[indices[i]];
+        }
+        return rs;
+    }
+
     private RecursiveTask<Integer> helperBAsync(final int goodFrom, final int goodUntil,
```

Comment:
getPoints and getRanks need a somewhat different treatment. We aim at zero GC overhead. The place for these points and ranks needs to be allocated in RedBlackTreeSweepHybridBOS, then the range in [from; until) needs to be filled.

Yet more, if you don't change the contents of the points returned from getPoints, you are safe to copy the array references, since it is an invariant throughout the library that you don't change the points.